21.md

```diff
@@ -0,0 +1,173 @@
+LUD-21: Signed LNURLs
+=====================================================
+
+`author: chill117`
+
+---
+
+## Authorize the creation of LNURLs for service
+
+A service may authorize other applications to create LNURLs without direct interaction between the authorized application and the service. This can allow offline hardware devices to generate LNURLs for a service that supports the signing scheme described in this document.
+
+The service must keep a list of authorized keys so that it can check whether a request was signed by an authorized key. The authorized keys list can be empty - in this case no signed requests will be accepted. An authorization key should include a secret key, the encoding of the secret key, and a unique identifier. An example authorization key as a JSON object:
+```json
+{
+	"id": "935e30a7",
+	"key": "e31b5c188346f3a83a7e698486bee48522eed378847126d78dbc030093ea14c7",
+	"encoding": "hex"
+}
+```
+Possible values for `"encoding"`:
+* `"base64"` - Base64 encoded
+* `"hex"` - Hexadecimal encoded
+* `""` (empty-string) - Unencoded / plaintext (utf8)
+
+The `"id"` should be a unique identifier so that an individual authorization key may be found by this value.
+
+An authorization key is to be provided to the authorized application so that it can generate signed URLs.
+
+
+### Steps to generate a signed URL
+
+These steps are to be done by an authorized application.
+
+1. __Build the base URL__ in the same way that your service would normally less the `k1` or secret value. For example, if your service generates withdraw links like https://example.com/lnurl?tag=withdraw&amount=5&currency=EUR&k1=0aa4a2285fb16207865c87c28bb0d78f42248f3077aba917c2f55b6be51e5e3d where `k1` is the secret that makes the link unique and acts as a secret (or password) which grants access to the URL. The `k1` can be left out here because we are using signing to authorize the creation of the URL in the server. So the example URL here would be as follows:
+	* https://example.com/lnurl?tag=withdraw&amount=5&currency=EUR
+2. Add the authorization key's identifier to the URL's query string as follows:
+	* https://example.com/lnurl?tag=withdraw&amount=5&currency=EUR&id=935e30a7
+3. Generate a unique, random nonce and add it to the query string:
```

Comment:
Unique is enough, doesn't need to be random. (And actually just incrementing a number by one and storing in persistent memory may be safer - no RNG vulnerabilities possible.)